### PR TITLE
Find Chrome binary on more platforms and fail gracefully

### LINKED ISSUE #10992: wpt run chrome does not use the correct default OSX binary location when getting version string
**What happens**

Fails out before launching the browser when I try to run wpt run on my mac.

**How to reproduce**

`./wpt run chrome infrastructure/testdriver/click.html  --log-mach-level=debug`
> Traceback (most recent call last):
  File "./wpt", line 5, in <module>
    wpt.main()
  File "/Users/jonathonkereliuk/0101/web-platform-tests/tools/wpt/wpt.py", line 132, in main
    rv = script(*args, **kwargs)
  File "/Users/jonathonkereliuk/0101/web-platform-tests/tools/wpt/run.py", line 437, in run
    **kwargs)
  File "/Users/jonathonkereliuk/0101/web-platform-tests/tools/wpt/run.py", line 424, in setup_wptrunner
    kwargs['browser_version'] = setup_cls.browser.version(kwargs.get("binary"))
  File "/Users/jonathonkereliuk/0101/web-platform-tests/tools/wpt/browser.py", line 342, in version
    version_string = call(binary, "--version").strip()
  File "/Users/jonathonkereliuk/0101/web-platform-tests/tools/wpt/utils.py", line 44, in call
    return subprocess.check_output(args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 566, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

It thinks that the binary should be at usr/bin/google-chrome.

I think this is low priority since the work around is explicitly specifying the binary location with --binary.

I believe this was introduced by #10935

### PR DESCRIPTION
**Pull request by Hexcles**
Also make similar changes to Opera.  Fixes #10992.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/11000)

### CODE REVIEW

File: tools/wpt/browser.py [hunk 1]
```diff
@@ -293,9 +293,17 @@ class Chrome(Browser):
...
-    binary = "/usr/bin/google-chrome"
     requirements = "requirements_chrome.txt"
 
+    @property
+    def binary(self):
+        if uname[0] == "Linux":
+            return "/usr/bin/google-chrome"
+        if uname[0] == "Darwin":
+            return "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
+        # TODO Windows?
```
Comment (kereliuk) [inline tools/wpt/browser.py]: Maybe we should have a debug log here that states that no matching uname was found so we didn't attempt to find the version binary?

File: tools/wpt/browser.py [hunk 2]
```diff
@@ -339,7 +347,10 @@ def install_webdriver(self, dest=None):
 
     def version(self, binary=None):
         binary = binary or self.binary
-        version_string = call(binary, "--version").strip()
+        try:
+            version_string = call(binary, "--version").strip()
+        except subprocess.CalledProcessError:
```
Comment (kereliuk) [inline tools/wpt/browser.py]: Same suggestion about a debug log here.

File: tools/wpt/browser.py [hunk 3]
```diff
@@ -379,9 +390,15 @@ class Opera(Browser):
...
 
     product = "opera"
-    binary = "/usr/bin/opera"
     requirements = "requirements_opera.txt"
 
+    @property
+    def binary(self):
+        if uname[0] == "Linux":
+            return "/usr/bin/opera"
+        # TODO Windows, Mac?
```
Comment (kereliuk) [inline tools/wpt/browser.py]: And here :)

File: tools/wpt/browser.py [hunk 4]
```diff
@@ -430,7 +447,10 @@ def install_webdriver(self, dest=None):
     def version(self, binary):
         """Retrieve the release version of the installed browser."""
         binary = binary or self.binary
-        output = call(binary, "--version")
+        try:
+            output = call(binary, "--version")
+        except subprocess.CalledProcessError:
```
Comment (kereliuk) [inline tools/wpt/browser.py]: Final spot!

**Review by kereliuk — APPROVED** [review]
Looks good! Just some nits about adding some debug logs for when we don't match the uname or the subprocess fails.

File: tools/wpt/browser.py [hunk 5]
```diff
@@ -293,9 +293,17 @@ class Chrome(Browser):
...
     product = "chrome"
-    binary = "/usr/bin/google-chrome"
     requirements = "requirements_chrome.txt"
 
+    @property
+    def binary(self):
+        if uname[0] == "Linux":
+            return "/usr/bin/google-chrome"
+        if uname[0] == "Darwin":
+            return "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
```
Comment (gsnedders) [inline tools/wpt/browser.py]: Technically, I don't think this is guaranteed. I think technically on macOS we should read Contents/Info.plist, read the CFBundleExecutable key, and then set the binary to Contents/MacOS/[that value]. That said, probably needless pedantry.

Comment (Hexcles) [reply tools/wpt/browser.py]: You're right; this is BAD(tm). Also the bundle might be installed to a user-specific applications directory. Not to mention I made zero effort for Windows...

I'm essentially using a simplified version of Chromedriver's heuristics: https://cs.chromium.org/chromium/src/chrome/test/chromedriver/chrome/chrome_finder.cc , which is also not ideal.

Basically the only reliable way is to always explicitly specify the --binary flag.

Comment (gsnedders) [reply tools/wpt/browser.py]: I at some point had some plan to implement some `find_macos_binary_path(bundle_id)`, which took the application id (e.g., `com.google.Chrome`) and then returns the path to the binary. Probably even splitting it up into two functions (one which finds the `.app` bundle, one which finds the binary within it).

Comment (Hexcles) [reply tools/wpt/browser.py]: Let's wish it'll happen some day :-P

**Pull request merged**